[SPARK-15689][SQL] data source v2 read path

cloud-fan opened the pull request:
## What changes were proposed in this pull request?

This PR adds the infrastructure for data source v2, and implement features which Spark already have in data source v1, i.e. column pruning, filter push down, catalyst expression filter push down, InternalRow scan, schema inference, data size report. The write path is excluded to avoid making this PR growing too big, and will be added in follow-up PR.

## How was this patch tested?

new tests

cloud-fan committed:
data source v2

cloud-fan commented:
cc @rxin @j-baker @rdblue

SparkQA commented:
**[Test build #81412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81412/testReport)** for PR 19136 at commit [`543a40b`](https://github.com/apache/spark/commit/543a40b2c9d4c456f8a726a12f78bd7f7d529b93).

SparkQA commented:
**[Test build #81412 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81412/testReport)** for PR 19136 at commit [`543a40b`](https://github.com/apache/spark/commit/543a40b2c9d4c456f8a726a12f78bd7f7d529b93).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `public abstract class DataSourceV2Reader `
  * `class RowToUnsafeRowReadTask implements ReadTask<UnsafeRow> `
  * `class RowToUnsafeDataReader implements DataReader<UnsafeRow> `
  * `class DataSourceRDDPartition[T : ClassTag](val index: Int, val readTask: ReadTask[T])`
  * `class DataSourceRDD[T: ClassTag](`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81412/
Test FAILed.

SparkQA commented:
**[Test build #81441 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81441/testReport)** for PR 19136 at commit [`a824d44`](https://github.com/apache/spark/commit/a824d44f9a4aac0518c5cd30893c34b36a094798).

SparkQA commented:
**[Test build #81441 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81441/testReport)** for PR 19136 at commit [`a824d44`](https://github.com/apache/spark/commit/a824d44f9a4aac0518c5cd30893c34b36a094798).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `public abstract class DataSourceV2Reader `
  * `class RowToUnsafeRowReadTask implements ReadTask<UnsafeRow> `
  * `class RowToUnsafeDataReader implements DataReader<UnsafeRow> `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81441/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #81466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81466/testReport)** for PR 19136 at commit [`a824d44`](https://github.com/apache/spark/commit/a824d44f9a4aac0518c5cd30893c34b36a094798).

RussellSpitzer commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/DataSourceV2Reader.java:
```diff
@@ -0,0 +1,126 @@
+ *   2. propagate information upward to Spark, e.g., report statistics, report ordering, etc.
+ * Spark first applies all operator push down optimizations which this data source supports. Then
+ * Spark collects information this data source provides for further optimizations. Finally Spark
+ * issues the scan request and does the actual data reading.
```
This would be really nice imho.

SparkQA commented:
**[Test build #81466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81466/testReport)** for PR 19136 at commit [`a824d44`](https://github.com/apache/spark/commit/a824d44f9a4aac0518c5cd30893c34b36a094798).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `public abstract class DataSourceV2Reader `
  * `class RowToUnsafeRowReadTask implements ReadTask<UnsafeRow> `
  * `class RowToUnsafeDataReader implements DataReader<UnsafeRow> `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81466/
Test PASSed.

rdblue commented:
Thanks for pinging me. I left comments on the older PR, since other discussion was already there. If you'd prefer comments here, just let me know.

SparkQA commented:
**[Test build #81490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81490/testReport)** for PR 19136 at commit [`89cbfb7`](https://github.com/apache/spark/commit/89cbfb7c98325852c5c97d321d83fe91154b129e).

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/SchemaRequiredDataSourceV2.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * A variant of `DataSourceV2` which requires users to provide a schema when reading data. A data
+ * source can inherit both `DataSourceV2` and `SchemaRequiredDataSourceV2` if it supports both schema
+ * inference and user-specified schemas.
```
cc @rdblue for the new API of schema reference.

sureshthalamati replied:
The check  added for addOption protects modifying the options passed to the datasource, but data source can still add new options by accident.  I think it might be safer to pass DataSourceV2Options that are **Unmodifiable** by the data source.

SparkQA commented:
**[Test build #81490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81490/testReport)** for PR 19136 at commit [`89cbfb7`](https://github.com/apache/spark/commit/89cbfb7c98325852c5c97d321d83fe91154b129e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `public abstract class DataSourceV2Reader `
  * `class RowToUnsafeRowReadTask implements ReadTask<UnsafeRow> `
  * `class RowToUnsafeDataReader implements DataReader<UnsafeRow> `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81490/
Test PASSed.

sureshthalamati commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/DataSourceV2Reader.java:
```diff
@@ -0,0 +1,126 @@
+   */
+  @Experimental
+  @InterfaceStability.Unstable
+  public List<ReadTask<UnsafeRow>> createUnsafeRowReadTasks() {
```
I really like the new API's flexibility to implement the different types of support.  Considering UnsafeRow is unstable , Would it be possible to move  createUnsafeRowReadTasks  to  a different interface ?  That might make data source implement two types of data sources  one with Row , and another one with UnsafeRow and make it easily configurable based on the spark version.

sureshthalamati commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java:
```diff
@@ -0,0 +1,56 @@
+   * Adds one more entry to the options.
+   * This should only be called by Spark, not data source implementations.
+   */
+  public void addOption(String key, String value) {
```
The check added for addOption protects modifying the options passed to the datasource, but data source can still add new options by accident. I think it might be safer to pass DataSourceV2Options that are Unmodifiable by the data source.

cloud-fan replied:
good point, I'll make it immutable.

SparkQA commented:
**[Test build #81506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81506/testReport)** for PR 19136 at commit [`ee5faf1`](https://github.com/apache/spark/commit/ee5faf109b279fe7ec03f7f1d98f6635a89060cb).

SparkQA commented:
**[Test build #81510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81510/testReport)** for PR 19136 at commit [`182b89d`](https://github.com/apache/spark/commit/182b89d1421c48049b7460e32586c90032e042c6).

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/scan/UnsafeRowScan.java:
```diff
@@ -0,0 +1,49 @@
+ */
+@Experimental
+@InterfaceStability.Unstable
+public interface UnsafeRowScan extends DataSourceV2Reader {
```
cc @j-baker for the new unsafe row scan API. Programmatically unsafe row scan should be in the base class, and normal row scan should be in the child class. However, conceptually for a developer, normal row scan is a basic interface and should be in the base class. Unsafe row scan is kind of an add-on and should be in the child class.

ash211 commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/upward/StatisticsSupport.java:
```diff
@@ -0,0 +1,26 @@
+ * A mix in interface for `DataSourceV2Reader`. Users can implement this interface to report
+ * statistics to Spark.
+ */
+public interface StatisticsSupport {
```
some datasources have per-column statistics, like how many bytes a column has or its min/max (e.g. things required for CBO).

should that be a separate interface from this one?

cloud-fan replied:
I'd like to put column stats in a separated interface, because we already separate basic stats and column stats in `ANALYZE TABLE`.

SparkQA commented:
**[Test build #81506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81506/testReport)** for PR 19136 at commit [`ee5faf1`](https://github.com/apache/spark/commit/ee5faf109b279fe7ec03f7f1d98f6635a89060cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`
  * `class RowToUnsafeRowReadTask(rowReadTask: ReadTask[Row], schema: StructType)`
  * `class RowToUnsafeDataReader(rowReader: DataReader[Row], encoder: ExpressionEncoder[Row])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81506/
Test PASSed.

SparkQA commented:
**[Test build #81510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81510/testReport)** for PR 19136 at commit [`182b89d`](https://github.com/apache/spark/commit/182b89d1421c48049b7460e32586c90032e042c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`
  * `class RowToUnsafeRowReadTask(rowReadTask: ReadTask[Row], schema: StructType)`
  * `class RowToUnsafeDataReader(rowReader: DataReader[Row], encoder: ExpressionEncoder[Row])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81510/
Test PASSed.

RussellSpitzer commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/ReadTask.java:
```diff
@@ -0,0 +1,43 @@
+   * task will always run on these locations. Implementations should make sure that it can
+   * be run on any location.
+   */
+  default String[] preferredLocations() {
```
These have previously only been ip/hostnames. To match the RDD definition I think we would have to continue with that.

cloud-fan replied:
This API matches the `RDD.preferredLocations` directly, I'll add more documents here.

j-baker replied:
can we have a class Host which represents this? Just makes the API more clear.

cloud-fan replied:
hmmm, do you mean create a `Host` class which only has a string field?

rdblue commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/SchemaRequiredDataSourceV2.java:
```diff
@@ -0,0 +1,42 @@
+   *
+   * @param schema the full schema of this data source reader. Full schema usually maps to the
+   *               physical schema of the underlying storage of this data source reader, e.g.
+   *               parquet files, JDBC tables, etc, while this reader may not read data with full
```
Maybe update the doc here, since JDBC sources and Parquet files probably shouldn't implement this. CSV and JSON are the examples that come to mind for sources that require a schema.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala:
```diff
@@ -0,0 +1,39 @@
+
+object DataSourceV2Relation {
+  def apply(reader: DataSourceV2Reader): DataSourceV2Relation = {
+    new DataSourceV2Relation(reader.readSchema().toAttributes, reader)
```
Is this the right schema? The docs for `readSchema` say it is the result of pushdown and projection, which doesn't seem appropriate for a `Relation`. Does relation represent a table that can be filtered and projected, or does it represent a single read? At least in the Hive read path, it's a table.

cloud-fan replied:
I leave it as a [TODO](https://github.com/apache/spark/pull/19136#discussion_r137023744) as it needs some refactoring on the optimizer. For now `DataSourceV2Relation` represents a data source without any optimization: we do these optimizations during planning. This is also a problem for data source v1, and that's why we implement partition pruning as an optimizer rule instead of data source internal, because we need to update the stats.

rdblue replied:
Are you saying that partition pruning isn't delegated to the data source in this interface?

I was just looking into how the data source should provide partition data, or at least fields that are the same for all rows in a `ReadTask`. It would be nice to have a way to pass those up instead of materializing them in each `UnsafeRow`.

cloud-fan replied:
In data source V2, we will delegate partition pruning to the data source, although we need to do some refactoring to make it happen.

> I was just looking into how the data source should provide partition data, or at least fields that are the same for all rows in a `ReadTask`. It would be nice to have a way to pass those up instead of materializing them in each `UnsafeRow`.

This can be achieved by the columnar reader. Think about a data source having a data column `i` and a partition column `j`, the returned columnar batch has 2 column vectors for `i` and `j`. Column vector `i` is a normal one that contains all the values of column `i` within this batch, column vector `j` is a constant vector that only contains a single value.

rdblue replied:
I think we should add a way to provide partition values outside of the columnar reader. It wouldn't be too difficult to add a method on `ReadTask` that returns them, then create a joined row in the scan exec. Otherwise, this requires a lot of wasted memory for a scan.

cloud-fan replied:
I think users can write a special `ReadTask` to do it, but we can't save memory by doing this. When an operator(the scan operator) transfers data to another operator, the data must be `UnsafeRow`s. So even users return a joined row in `ReadTask`, Spark need to convert it to `UnsafeRow`.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:
```diff
@@ -0,0 +1,95 @@
+        case r: CatalystFilterPushDownSupport =>
+          r.pushCatalystFilters(filters.toArray)
+
+        case r: FilterPushDownSupport =>
```
Looks like `CatalystFilterPushDownSupport` and `FilterPushDownSupport` are exclusive?

cloud-fan replied:
yea we can't prevent users to implement them both, and we will pick `CatalystFilterPushDownSupport` over `FilterPushDownSupport`. Let me document it.

j-baker replied:
Considering that there is a translation between Catalyst filters and Filters, it's probably worth _just_ doing the catalyst one, and providing the user with the translator if they want to do the Filter approach?

j-baker replied:
can FilterPushDownSupport be an interface which extends CatalystFilterPushDownSupport and provides a default impl of pruning the catalyst flter? Like, this code can just go there as a method:

```
interface FilterPushDownSupport extends CatalystFilterPushDownSupport {
    List<Filter> pushFilters(List<Filter> filters);

    default List<Expression> pushCatalystFilters(List<Expression> filters) {
        Map<Filter, Expression> translatedMap = new HashMap<>();
        List<Filter> nonconvertiblePredicates = new ArrayList<>();

        for (Expression catalystFilter : filters) {
            Optional<Filter> translatedFilter = DataSourceStrategy.translateFilter(catalystFilter);
            if (translatedFilter.isPresent()) {
                translatedMap.put(translatedFilter.get(), catalystFilter);
            } else {
                nonconvertiblePredicates.add(catalystFilter);
            }
        }

        List<Filter> unhandledFilters = pushFilters(new ArrayList<>(translatedMap.values()));
        return Stream.concat(
            nonconvertiblePredicates.stream(),
            unhandledFilters().stream().map(translatedMap::get))
           .collect(toList());
    }
}
```

and we can trivially ignore the interface confusion (it's truly confusing if you can implement two interfaces)

j-baker replied:
like, we might as well not document it if the code can document it

cloud-fan replied:
good idea!

viirya replied:
By doing so, do we still need to match both `CatalystFilterPushDownSupport` and `FilterPushDownSupport` here?

cloud-fan replied:
After some attempts, I went back with 2 individual interfaces. The reason is that, a) `CatalystFilterPushDownSupport` is an unstable interface, and it looks weird to let a stable interface extend an unstable one. b) the logic that converts expressions to public filters belongs to Spark internal, and we may change it in the future, so we should not put these codes in a public interface. We may have a risk of breaking compatibility for this interface.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:
```diff
@@ -0,0 +1,95 @@
+
+      // Match original case of attributes.
+      // TODO: nested fields pruning
+      val requiredColumns = (projectSet ++ filterSet).toSeq.map(attrMap)
```
Do we need to request columns that are only referenced by pushed filters?

rdblue replied:
It seems reasonable to only request the ones that will be used, or that have residuals after pushing filters.

gatorsmile commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/upward/StatisticsSupport.java:
```diff
@@ -0,0 +1,26 @@
+ * statistics to Spark.
+ */
+public interface StatisticsSupport {
+  Statistics getStatistics();
```
Will the returned stats be adjusted by the data sources based on the operator push-down?

cloud-fan replied:
It should, but we need some refactor on optimizer, see https://github.com/apache/spark/pull/19136#discussion_r137023744

j-baker commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/downward/CatalystFilterPushDownSupport.java:
```diff
@@ -0,0 +1,36 @@
+  /**
+   * Push down filters, returns unsupported filters.
+   */
+  Expression[] pushCatalystFilters(Expression[] filters);
```
any chance this could push java lists? They're just more idiomatic in a java interface

cloud-fan replied:
java list is not friendly to scala implementations :)

j-baker commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/upward/Statistics.java:
```diff
@@ -0,0 +1,28 @@
+ * An interface to represent statistics for a data source.
+ */
+public interface Statistics {
+  long sizeInBytes();
```
OptionalLong for sizeInBytes? It's not obvious that sizeInBytes is well defined for e.g. JDBC datasources, but row count can generally be easily estimated from the query plan.

j-baker replied:
like, I get that it's non-optional at the moment, but it's odd that we have a method that the normal implementor will have to replace with

```
public long sizeInBytes() {
    return Long.MAX_VALUE;
}
```

j-baker replied:
and now is a good time to fix it :)

j-baker commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java:
```diff
@@ -0,0 +1,57 @@
+
+  /**
+   * Returns the option value to which the specified key is mapped, case-insensitively,
+   * or {@code null} if there is no mapping for the key.
```
can we return `Optional<String>` here? JDK maintainers wish they could return optional on Map

j-baker commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java:
```diff
@@ -0,0 +1,57 @@
+   * Returns the option value to which the specified key is mapped, case-insensitively,
+   * or {@code defaultValue} if there is no mapping for the key.
+   */
+  public String getOrDefault(String key, String defaultValue) {
```
if the above returns `Optional`, you probably don't need this method.

j-baker commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java:
```diff
@@ -0,0 +1,57 @@
+ * options.
+ */
+public class DataSourceV2Options {
+  private Map<String, String> keyLowerCasedMap;
```
nit: final

cloud-fan commented:
I think I've addressed all the comments, but I still have some naming that I'm not confident about.

1. is it OK to add the `V2` suffix in all related classes and packages?
2. I created 3 packages for the 3 kinds of optimization traits: `downward`, `upward` and `scan`. Any better ideas?

SparkQA commented:
**[Test build #81712 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81712/testReport)** for PR 19136 at commit [`65e994d`](https://github.com/apache/spark/commit/65e994da5ad02b83a71dd77bcbfe0716b627c99c).

SparkQA commented:
**[Test build #81712 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81712/testReport)** for PR 19136 at commit [`65e994d`](https://github.com/apache/spark/commit/65e994da5ad02b83a71dd77bcbfe0716b627c99c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`
  * `class RowToUnsafeRowReadTask(rowReadTask: ReadTask[Row], schema: StructType)`
  * `class RowToUnsafeDataReader(rowReader: DataReader[Row], encoder: ExpressionEncoder[Row])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81712/
Test FAILed.

SparkQA commented:
**[Test build #81717 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81717/testReport)** for PR 19136 at commit [`4ff1b18`](https://github.com/apache/spark/commit/4ff1b18d3db9f50ba7f3d31288d0da37736d6b5f).

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java:
```diff
@@ -0,0 +1,49 @@
+ * An immutable case-insensitive string-to-string map, which is used to represent data source
+ * options.
+ */
+public class DataSourceV2Options {
```
add a simple test suite for this

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2Options.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Optional;
+
+/**
+ * An immutable case-insensitive string-to-string map, which is used to represent data source
```
we need to be clear that only the keys are case insensitive. the values are case preserving.

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/SchemaRequiredDataSourceV2.java:
```diff
@@ -0,0 +1,42 @@
+ * source can inherit both `DataSourceV2` and `SchemaRequiredDataSourceV2` if it supports both schema
+ * inference and user-specified schemas.
+ */
+public interface SchemaRequiredDataSourceV2 {
```
what's an example of such data source?

rxin replied:
I personally find this divergence at the top pretty confusing ...

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/DataReader.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * A data reader returned by a read task and is responsible for outputting data for a RDD partition.
+ */
+public interface DataReader<T> extends Closeable {
```
what can T be?

cloud-fan replied:
currently it can be `Row`, `UnsafeRow`, `ColumnarBatch`.

rxin replied:
Document this and link it back to whatever method it is.

Also I'd still add an explicit init or open.

rdblue replied:
The initialization is done when creating this `DataReader` from a `ReadTask`. That ensures that the initialization happens (easy to forget `open()`) and simplifies the checks that need to be done because `DataReader` can't exist otherwise.

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/downward/ColumnPruningSupport.java:
```diff
@@ -0,0 +1,36 @@
+   * also OK to do the pruning partially, e.g., a data source may not be able to prune nested
+   * fields, and only prune top-level columns.
+   */
+  void pruneColumns(StructType requiredSchema);
```
link this to readSchema function

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/upward/Statistics.java:
```diff
@@ -0,0 +1,28 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader.upward;
```
this package name is really confusing. maybe just put all of them in the v2.reader package. There isn't that many classes ... if you are worried about discoverability, use a common interface, or create a top level class and put the interfaces there.

yueawang commented:
@cloud-fan, Last week when I saw this PR, and I remember that you also implemented some new pushdowns like sort or limit, are they removed in this the latest commit?? Any concern? Thanks!

cloud-fan committed:
naming updates

SparkQA commented:
**[Test build #81722 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81722/testReport)** for PR 19136 at commit [`1e86d5c`](https://github.com/apache/spark/commit/1e86d5ca445d732af6ac651d49d391d5cd012a92).

cloud-fan commented:
@yueawang these new push-downs are in my [prototype](https://github.com/cloud-fan/spark/pull/10). This PR is the first version of data source v2, so I'd like to cut down the patch size and only implement features that we already have in data source v1.

SparkQA commented:
**[Test build #81723 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81723/testReport)** for PR 19136 at commit [`abcc606`](https://github.com/apache/spark/commit/abcc606e006e9975d1507eed379a48a3134165ad).

SparkQA commented:
**[Test build #81717 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81717/testReport)** for PR 19136 at commit [`4ff1b18`](https://github.com/apache/spark/commit/4ff1b18d3db9f50ba7f3d31288d0da37736d6b5f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class DataSourceV2Options `
  * `class DataSourceRDDPartition(val index: Int, val readTask: ReadTask[UnsafeRow])`
  * `class DataSourceRDD(`
  * `case class DataSourceV2Relation(`
  * `case class DataSourceV2ScanExec(`
  * `class RowToUnsafeRowReadTask(rowReadTask: ReadTask[Row], schema: StructType)`
  * `class RowToUnsafeDataReader(rowReader: DataReader[Row], encoder: ExpressionEncoder[Row])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81717/
Test PASSed.

SparkQA commented:
**[Test build #81722 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81722/testReport)** for PR 19136 at commit [`1e86d5c`](https://github.com/apache/spark/commit/1e86d5ca445d732af6ac651d49d391d5cd012a92).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81722/
Test PASSed.

SparkQA commented:
**[Test build #81723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81723/testReport)** for PR 19136 at commit [`abcc606`](https://github.com/apache/spark/commit/abcc606e006e9975d1507eed379a48a3134165ad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81723/
Test PASSed.

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Statistics.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.OptionalLong;
+
+/**
+ * An interface to represent statistics for a data source.
```
link back to SupportsReportStatistics

cloud-fan committed:
more comments

SparkQA commented:
**[Test build #81773 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81773/testReport)** for PR 19136 at commit [`a1301f5`](https://github.com/apache/spark/commit/a1301f5d86847a06adc872cad9a93a28884de3fe).

SparkQA commented:
**[Test build #81773 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81773/testReport)** for PR 19136 at commit [`a1301f5`](https://github.com/apache/spark/commit/a1301f5d86847a06adc872cad9a93a28884de3fe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81773/
Test PASSed.

cloud-fan commented:
any more comments? is it ready to go?

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2.java:
```diff
@@ -0,0 +1,28 @@
+ * constructor.
+ *
+ * Note that this is an empty interface, data source implementations should mix-in at least one of
+ * the plug-in interfaces like `ReadSupport`. Otherwise it's just a dummy data source which is
```
use an actual link ...

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/ReadSupport.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.spark.sql.sources.v2.reader.DataSourceV2Reader;
+
+/**
+ * A mix-in interface for `DataSourceV2`. Users can implement this interface to provide data reading
```
Users -> data source implementers

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/ReadSupportWithSchema.java:
```diff
@@ -0,0 +1,45 @@
+ * source can implement both `ReadSupport` and `ReadSupportWithSchema` if it supports both schema
+ * inference and user-specified schema.
+ */
+public interface ReadSupportWithSchema {
```
I still find ReadSupport vs ReadSupportWithSchema pretty confusing. But let's address that separately.

rxin commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Statistics.java:
```diff
@@ -0,0 +1,29 @@
+
+/**
+ * An interface to represent statistics for a data source, which is returned by
+ * `SupportsReportStatistics`.
```
also use `@link`

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala:
```diff
@@ -0,0 +1,71 @@
+
+class DataSourceRDD(
+    sc: SparkContext,
+    @transient private val generators: java.util.List[ReadTask[UnsafeRow]])
```
why is this called a generators?

rxin commented:
LGTM. 

Still some feedback that can be addressed later. We should also document all the APIs as Evolving.

cloud-fan committed:
document improvement

SparkQA commented:
**[Test build #81809 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81809/testReport)** for PR 19136 at commit [`d2c86f4`](https://github.com/apache/spark/commit/d2c86f4339d59f227bef61b1c97f9770ce1233b9).

SparkQA commented:
**[Test build #81809 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81809/testReport)** for PR 19136 at commit [`d2c86f4`](https://github.com/apache/spark/commit/d2c86f4339d59f227bef61b1c97f9770ce1233b9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81809/
Test PASSed.

cloud-fan commented:
thank you all for the review, merging to master!
